cmd/util/ledger/migrations/account_based_migration.go

```diff
@@ -1,190 +1,366 @@
+					if job.Address == common.ZeroAddress {
+						resultCh <- &migrationResult{
+							migrationDuration: migrationDuration{
+								Address:  job.Address,
+								Duration: time.Since(start),
+							},
+							Migrated: job.Payloads,
+						}
+						continue
+					}
+
+					if _, ok := knownProblematicAccounts[job.Address]; ok {
+						log.Info().
+							Hex("address", job.Address[:]).
+							Msg("skipping problematic account")
+						resultCh <- &migrationResult{
+							migrationDuration: migrationDuration{
+								Address:  job.Address,
+								Duration: time.Since(start),
+							},
+							Migrated: job.Payloads,
+						}
+						continue
+					}
+
+					var err error
+					accountMigrated := job.Payloads
+					for m, migrator := range migrations {
+
+						select {
+						case <-ctx.Done():
+							return
+						default:
+						}
+
+						accountMigrated, err = migrator.MigrateAccount(ctx, job.Address, accountMigrated)
+						if err != nil {
+							log.Error().
+								Err(err).
+								Int("migration_index", m).
+								Type("migration", migrator).
+								Hex("address", job.Address[:]).
+								Msg("could not migrate account")
+							cancel(fmt.Errorf("could not migrate account: %w", err))
+							return
+						}
+					}
+
+					resultCh <- &migrationResult{
+						migrationDuration: migrationDuration{
+							Address:  job.Address,
+							Duration: time.Since(start),
+						},
+						Migrated: accountMigrated,
+					}
+				}
+			}
+		}()
 	}
 
-	log.Info().Msgf("finished migrating payloads for %v account", len(groups.Accounts))
+	go func() {
+		for {
+			g, err := accountGroups.Next()
+			if err != nil {
+				cancel(fmt.Errorf("could not get next account group: %w", err))
+				return
+			}
 
-	// add the non accounts which don't need to be migrated
-	migrated = append(migrated, groups.NonAccountPayloads...)
+			if g == nil {
+				break
+			}
 
-	log.Info().Msgf("finished migrating all account based payloads, total migrated payloads: %v", len(migrated))
+			job := jobMigrateAccountGroup{
+				Address:  g.Address,
+				Payloads: g.Payloads,
+			}
 
-	return migrated, nil
-}
+			select {
+			case <-ctx.Done():
+				return
+			case jobs <- job:
+			}
+		}
+	}()
 
-// MigrateGroupSequentially migrate the payloads in the given payloadsByAccount map which
-// using the migrator
-func MigrateGroupSequentially(
-	migrator AccountMigrator,
-	payloadsByAccount map[string][]ledger.Payload,
-) (
-	[]ledger.Payload, error) {
+	// read job results
+	logAccount := moduleUtil.LogProgress("processing account group", accountGroups.Len(), log)
+
+	migrated := make([]*ledger.Payload, 0)
+	durations := newMigrationDurations(logTopNDurations)
+	for i := 0; i < accountGroups.Len(); i++ {
+		select {
+		case <-ctx.Done():
+			break
+		case result := <-resultCh:
+			durations.Add(result)
+
+			accountMigrated := result.Migrated
+			migrated = append(migrated, accountMigrated...)
+			logAccount(1)
+		}
+	}
```

Comment:
I think we want to break out of the `accountGroups` loop instead of breaking out of `select`.

```suggestion
        accountGroupLoop:
	for i := 0; i < accountGroups.Len(); i++ {
		select {
		case <-ctx.Done():
			break accountGroupLoop
		case result := <-resultCh:
			durations.Add(result)

			accountMigrated := result.Migrated
			migrated = append(migrated, accountMigrated...)
			logAccount(1)
		}
	}
```